> Wasn't able to get this working so I can test it. After running initial calibration, I got the following error:
> 
> ```
> Exception in thread Thread-1:
> Traceback (most recent call last):
>   File "/Users/zacowan/.pyenv/versions/3.9.13/lib/python3.9/threading.py", line 980, in _bootstrap_inner
>     self.run()
>   File "/Users/zacowan/.pyenv/versions/3.9.13/lib/python3.9/threading.py", line 917, in run
>     self._target(*self._args, **self._kwargs)
>   File "/Users/zacowan/Development/Senior-Project-Fall-2022/IrisSoftware/main.py", line 242, in processing
>     topLeft, bottomRight = self.interpolator.getLeftEyeBox()
> TypeError: cannot unpack non-iterable NoneType object
> ```

That's the code that draws the boxes around your eyes in the live preview. Ah, I see the issue. The only way that error could happen is if `self.state.interpolationType` is set to `InterpolationType.JOYSTICK`, but somehow `InterpolationType.JOYSTICK` never got passed to `interpolator.calibrateInterpolator()`. The variable gets passed in correctly in the `IrisSoftware` constructor, but somehow there's another call to `calibrateInterpolator()` in the `run()` method that never passes in the interpolator type. It looks like that code only runs when you launch the program with no calibration data saved already. I pushed a fix. 

While I was at it, I noticed that the program never re-calibrates the interpolator if the user clicks "calibrate" while the program is running. I pushed a fix for that too.

--- a/IrisSoftware/computeScreenCoords.py
+++ b/IrisSoftware/computeScreenCoords.py
@@ -171,8 +171,7 @@ class JoystickInterpolator():
             # print('Moving down')
             return pyautogui.position()[0], self.screenYMax
 
-    # what if both at same time? Make nested
-    # move laterally, not in a diamond
+    # what if both at same time? Make nested?
 
     def getLeftEyeBox(self):
         return (round(self.leftXMin), round(self.leftYMin)), (round(self.leftXMax), round(self.leftYMax))

--- a/IrisSoftware/detectEyes.py
+++ b/IrisSoftware/detectEyes.py
@@ -128,13 +128,13 @@ class EyeDetection:
     def eyeCascadeDetector(self, image):
         '''Detect eyes using a single Haar cascade detector'''
         gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
-        return self.eyeDetector.detectMultiScale(gray)
+        return sorted(self.eyeDetector.detectMultiScale(gray), key=lambda eye: eye[0])
 
     def eyeCascadeBlobDetector(self, image):
         '''Detect eyes using a Haar cascade detector and blob detection'''
         gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
 
-        eyes = self.eyeDetector.detectMultiScale(gray)
+        eyes = sorted(self.eyeDetector.detectMultiScale(gray), key = lambda eye: eye[0])
         croppedEyes = [gray[y:y+h, x:x+w] for (x, y, w, h) in eyes]
 
         eyes_bw = [cv2.threshold(eye, self.blobThreshold, 255, cv2.THRESH_BINARY)[1]
@@ -165,7 +165,7 @@ class EyeDetection:
         faceX, faceY, faceW, faceH = face
         croppedFace = gray[faceY:faceY+faceH, faceX:faceX+faceW]
 
-        eyes = self.eyeDetector.detectMultiScale(croppedFace)
+        eyes = sorted(self.eyeDetector.detectMultiScale(croppedFace), key = lambda eye: eye[0])
 
         return [tupleAdd((x+w/2, y+h/2), faceX, faceY) for (x, y, w, h) in eyes]
 
@@ -193,7 +193,7 @@ class EyeDetection:
             cv2.waitKey()
             cv2.imwrite(f"image3_croppedFace.jpg", croppedFace)
 
-        eyes = self.eyeDetector.detectMultiScale(croppedFace)
+        eyes = sorted(self.eyeDetector.detectMultiScale(croppedFace), key = lambda eye: eye[0])
         croppedEyes = [croppedFace[y:y+h, x:x+w] for (x, y, w, h) in eyes]
         if demo:
             for i, eye in enumerate(croppedEyes):
@@ -250,9 +250,11 @@ def testRealtimeEyeDetection():
         # if len(eyes) == 0:
         #     print("No eyes detected")
 
-        for eye in eyes:
-            if len(eye) == 2:
-                cv2.circle(image, (eye[0], eye[1]), 7, (0, 0, 255), 2)
+        for i, eye in enumerate(eyes):
+            if i == 0:
+                cv2.circle(image, (eye[0], eye[1]), 7, (255, 0, 0), 2)
+            elif i == 1:
+                cv2.circle(image, (eye[0], eye[1]), 7, (0, 255, 0), 2)
         cv2.imshow("Eyes", image)
         if cv2.waitKey(delay=1) & 0xFF == ord('q'):
             break

--- a/IrisSoftware/main.py
+++ b/IrisSoftware/main.py
@@ -21,6 +21,7 @@ class IrisSoftware:
         def __init__(self) -> None:
             self.shouldExit = False
             self.isCalibrated = False
+            self.currentlyCalibrating = False
             self.calibrationEyeCoords: list[list[tuple]] = []
             self.faceBoxes = []
             self.faceBox = None
@@ -114,6 +115,7 @@ class IrisSoftware:
 
     def resetCalibrationEyeCoords(self):
         self.state.calibrationEyeCoords = []
+        self.state.currentlyCalibrating = False
         print("Reset current calibration eye coords.")
 
     def performInitialConfiguration(self):
@@ -148,6 +150,9 @@ class IrisSoftware:
 
     def captureCalibrationEyeCoords(self):
         """Captures and stores a eye coords for calibration."""
+        # not best spot to set currently calibrating to true, ideally should be another signal that is called under __openCalibration in ui.py or in calibrationwindow widget initialization
+        self.state.currentlyCalibrating = True
+
         eyeCoords = []
         maxFramesToCapture = 30
 
@@ -213,15 +218,18 @@ class IrisSoftware:
 
         print("Saved new calibration data.")
 
-        # Reset current calibration frames
-        self.resetCalibrationEyeCoords()
-
         # Train screen coords model
         self.interpolator.calibrateInterpolator(CALIBRATION_FILE_NAME, self.state.interpolatorType)
+
+        # Reset current calibration frames
+        self.resetCalibrationEyeCoords()
 
     def processing(self):
         """Thread to run main loop of eye detection"""
         while not self.state.shouldExit:
+            if self.state.currentlyCalibrating:
+                continue
+            
             # Get the camera frame
             frame = self.camera.getFrame()
 

